File: src/main/java/org/taktik/connector/technical/handler/CertificateCallback.java
package org.taktik.connector.technical.handler;

import org.taktik.connector.technical.exception.TechnicalConnectorException;
import org.taktik.connector.technical.exception.TechnicalConnectorExceptionValues;
import org.taktik.connector.technical.service.sts.security.Credential;
import org.taktik.connector.technical.service.sts.security.impl.KeyPairCredential;
import org.taktik.connector.technical.session.Session;
import java.security.PrivateKey;
import java.security.cert.X509Certificate;
import java.util.concurrent.TimeUnit;
import javax.xml.ws.handler.soap.SOAPMessageContext;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class CertificateCallback extends AbstractWsSecurityHandler {
   private static final Logger LOG = LoggerFactory.getLogger(CertificateCallback.class);
   private Credential cred;

   public CertificateCallback() throws TechnicalConnectorException {
   }

   public CertificateCallback(X509Certificate certificate, PrivateKey privateKey) throws TechnicalConnectorException {
      this.cred = new KeyPairCredential(privateKey, certificate);
   }

   public CertificateCallback(Credential cred) throws TechnicalConnectorException {
      this.cred = cred;
   }

   protected void addWSSecurity(SOAPMessageContext context) throws TechnicalConnectorException {
      Credential lazyCred = this.cred;
      if (lazyCred == null) {
         LOG.debug("[Lazy Loading] Trying to load SAMLToken from session.");
         lazyCred = Session.getInstance().getSession().getHolderOfKeyCredential();
         if (lazyCred == null) {
            throw new TechnicalConnectorException(TechnicalConnectorExceptionValues.ERROR_TECHNICAL, new Object[]{"Unable to lazy load credential."});
         }
      }

      this.buildSignature().on(context.getMessage()).withTimeStamp(60L, TimeUnit.SECONDS).withBinarySecurityToken(lazyCred).sign(new AbstractWsSecurityHandler.SignedParts[]{AbstractWsSecurityHandler.SignedParts.BODY, AbstractWsSecurityHandler.SignedParts.TIMESTAMP, AbstractWsSecurityHandler.SignedParts.BST});
   }

   protected Logger getLogger() {
      return LOG;
   }
}

File: src/main/java/org/taktik/connector/technical/handler/CertificateCallback.kt
package org.taktik.connector.technical.handler

import org.apache.commons.logging.Log
import org.apache.commons.logging.LogFactory
import org.taktik.connector.technical.exception.TechnicalConnectorException
import org.taktik.connector.technical.service.sts.security.Credential
import org.taktik.connector.technical.service.sts.security.impl.KeyPairCredential
import java.security.PrivateKey
import java.security.cert.X509Certificate
import java.util.concurrent.TimeUnit
import javax.xml.ws.handler.soap.SOAPMessageContext

class CertificateCallback : AbstractWsSecurityHandler {
    private var cred: Credential? = null

    @Throws(TechnicalConnectorException::class)
    constructor(certificate: X509Certificate, privateKey: PrivateKey) {
        this.cred = KeyPairCredential(privateKey, certificate)
    }

    @Throws(TechnicalConnectorException::class)
    constructor(cred: Credential) {
        this.cred = cred
    }

    @Throws(TechnicalConnectorException::class)
    override fun addWSSecurity(context: SOAPMessageContext) {

        this.buildSignature().on(context.message).withTimeStamp(this.getTimeStampTTL(), TimeUnit.SECONDS).withBinarySecurityToken(cred)
            .sign(AbstractWsSecurityHandler.SignedParts.BODY, AbstractWsSecurityHandler.SignedParts.TIMESTAMP, AbstractWsSecurityHandler.SignedParts.BST)
    }

    override fun getLogger(): Log {
        return LOG
    }

    companion object {
        private val LOG = LogFactory.getLog(this::class.java)
    }
}
